File: lecture05/subsets.cpp
// subsets.cpp

#include <iostream>
#include <numeric>
#include <vector>

using namespace std;

// Functions

size_t subsets(vector<int> &s, vector<int> &c, size_t k) {
    if (k == c.size()) { // Base: have complete subset
    	// for (auto e : s) { cout << " " << e; }; cout << endl;
    	return (accumulate(s.begin(), s.end(), 0) % 3 == 0) ? 1 : 0;
    }

    // Recurse: skip current
    size_t count1 = subsets(s, c, k + 1);

    // Recurse: with current
    s.push_back(c[k]);
    size_t count2 = subsets(s, c, k + 1);
    s.pop_back(); // Reset subset
    
    return count1 + count2;
}

// Main execution

int main(int argc, char *argv[]) {
    vector<int> subset;
    vector<int> numbers = {0, 1, 2, 3, 4, 5, 6, 7, 8, 9};
    size_t      count   = subsets(subset, numbers, 0);

    cout << count << endl;
    return 0;
}

File: lecture05/subsets.py
#!/usr/bin/env python3

from typing import Any, Iterator

import itertools

# Constants

NUMBERS = list(range(0, 10))

# Functions

def subsets(s: list[Any], c: list[Any], k: int=0) -> Iterator[list[Any]]:
    '''
    s:  current subset
    c:  all candidate elements
    k:  index of current candidate element
    '''

    # Base case: we have exhausted all candidates
    if k == len(c):
        yield s

    # Recursive cases
    else:
        # Recurse without using current element
        yield from subsets(s, c, k + 1)

        # Recurse while using current element
        yield from subsets(s + [c[k]], c, k + 1)

# Main Execution

def main() -> None:
    count = 0
    for length in range(0, len(NUMBERS) + 1):
        for subset in itertools.combinations(NUMBERS, length):
            if sum(subset) % 3 == 0:
                count += 1

    print(count)

    count = sum(1 for subset in subsets([], NUMBERS) if sum(subset) % 3 == 0)
    print(count)

if __name__ == '__main__':
    main()
